openfisca_us_data/datasets/ce/ce.py

```diff
@@ -0,0 +1,279 @@
+import numpy as np
+
+import h5py
+from pandas import DataFrame, concat
+
+from openfisca_us_data.utils import US, dataset
+from openfisca_us_data.datasets.ce.raw_ce import RawCE
+
+
+@dataset
+class CE:
+    name = "ce"
+    model = US
+    # Ratio of mean cash contributions to charity vs total cash contribs, 2016
+    # www.bls.gov/opub/btn/volume-8/the-relationship-between-cash-contributions-pretax-income-and-age.htm
+    proportion_cash_contrib_to_charity = 396.59 / 2080.85
+
+    def generate(year: int) -> None:
+        """Saves organized Consumer Expenditure data in HDF5 format via h5py
+
+        Computes "months in scope" from existing variables, and uses it to
+        adjust weights and compute annual estimates. Finally, those estimates
+        to compute estimated carbon emissions per household.
+
+        Args:
+            year (int): The year of the Consumer Expenditure to use.
+        """
+        year = int(year)
+        if year not in RawCE.years:
+            RawCE.generate(year)
+
+        raw_data = RawCE.load(year)
+        ce = h5py.File(CE.file(year), mode="w")
+        ce.create_group("/household")  # Household quarterly data
+        ce.create_group("/annual")  # Annual estimates
+
+        # Concatenate 5 "quarters" of fmli data, add months in scope ----------
+        df_list = []
+        for quarter_data in raw_data.keys():
+            df_list.append(raw_data[quarter_data])
+
+        fmli_df = concat(df_list)
+        mis = fmli_df.apply(
+            lambda row: months_in_scope(row["interview_mo"], row["nominal_quarter"]),
+            axis=1,
+        )
+        fmli_df.insert(7, "months_in_scope", mis)
+        fmli_df = fmli_df.sort_values(["cu_id", "interview_id"])
+
+        # Add household variables to H5 File ----------------------------------
+        add_survey_vars(ce, fmli_df)
+        add_demographics(ce, fmli_df)
+        add_expenditures(ce, fmli_df)
+        add_carbon_emissions(ce)
+
+        # Add annual estimates to H5 File -------------------------------------
+        estimate_annual_quantity(
+            ce, "/household/demographics/income_before_tax", "demographics"
+        )
+        estimate_annual_quantity(ce, "/household/expenditures/alcohol")
+        estimate_annual_quantity(ce, "/household/emissions/co2_kg")
+        raw_data.close()
+        ce.close()
+
+
+def months_in_scope(interview_mo: int, nominal_quarter: int):
+    """Get the number of calendar months representing a nominal quarter
+
+    Args:
+        interview_mo (int): the calendar month (e.g., 2 for February) that
+            the Consumer Expenditure survey was taken
+        nominal_quarter (int): The CE quarter of 1-5, with 5 representing the
+            first quarter of the next year.
+
+    Returns:
+        months_in_scope (int): the number of calendar months (out of a
+            possible 3) that will represent the nominal quarter
+    """
+    months_in_scope = np.nan
```

Comment:
Don't think we need to initialize since all the cases are either provided or errors right?

Reply:
you're right, we don't.